// tyronzil-js/dist/src/lib/decentralized-identity/did-operations/did-deactivate.js
"use strict";
/*
    TyronZIL-js: Decentralized identity client for the Zilliqa blockchain platform
    Copyright (C) 2020 Julio Cesar Cabrapan Duarte

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const zcrypto = __importStar(require("@zilliqa-js/crypto"));
const sidetree_1 = require("../protocols/sidetree");
/** Generates a `Tyron DID-Deactivate` operation */
class DidDeactivate {
    /***            ****            ***/
    constructor(operation) {
        this.type = sidetree_1.OperationType.Deactivate;
        this.decentralized_identifier = operation.did;
        this.signature = "0x" + operation.signature;
    }
    /** Generates a Sidetree-based `DID-Deactivate` operation */
    static execute(input) {
        return __awaiter(this, void 0, void 0, function* () {
            const TYRON_HASH = input.state.tyron_hash.substring(2);
            const PREVIOUS_RECOVERY_KEY = zcrypto.getPubKeyFromPrivateKey(input.recoveryPrivateKey);
            const SIGNATURE = zcrypto.sign(Buffer.from(TYRON_HASH, 'hex'), input.recoveryPrivateKey, PREVIOUS_RECOVERY_KEY);
            /** Output data from a Tyron `DID-Deactivate` operation */
            const OPERATION_OUTPUT = {
                did: input.state.decentralized_identifier,
                signature: SIGNATURE
            };
            return new DidDeactivate(OPERATION_OUTPUT);
        });
    }
}
exports.default = DidDeactivate;
