// tyronzil-js/dist/src/lib/decentralized-identity/sidetree-protocol/models/document-model.js
"use strict";
/*
    TyronZIL-js: Decentralized identity client for the Zilliqa blockchain platform
    Copyright (C) 2020 Julio Cesar Cabrapan Duarte

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.PatchAction = void 0;
var PatchAction;
(function (PatchAction) {
    PatchAction["AddKeys"] = "add-public-keys";
    PatchAction["RemoveKeys"] = "remove-public-keys";
    PatchAction["AddServices"] = "add-service-endpoints";
    PatchAction["RemoveServices"] = "remove-service-endpoints";
    // Format of an additional custom action
    PatchAction["CustomAction"] = "-custom-action";
})(PatchAction = exports.PatchAction || (exports.PatchAction = {}));
