// tyronzil-sdk/dist/decentralized-identity/protocols/models/verification-method-models.js
"use strict";
/*
    tyronzil-sdk: Tyron DID SDK - Zilliqa's DID-Method at www.tyronZIL.com
    Copyright (C) 2020 Julio Cesar Cabrapan Duarte

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicKeyPurpose = void 0;
var PublicKeyPurpose;
(function (PublicKeyPurpose) {
    PublicKeyPurpose["General"] = "general";
    PublicKeyPurpose["Auth"] = "authentication";
    PublicKeyPurpose["Assertion"] = "assertion";
    PublicKeyPurpose["Agreement"] = "agreement";
    PublicKeyPurpose["Invocation"] = "invocation";
    PublicKeyPurpose["Delegation"] = "delegation";
    PublicKeyPurpose["XSGD"] = "xsgd";
})(PublicKeyPurpose = exports.PublicKeyPurpose || (exports.PublicKeyPurpose = {}));
